### PR TITLE
.github, caplin: harden Gloas genesis verification

### PR DESCRIPTION
**Pull request by domiwei**
Follow-up to #23889, addressing the late review feedback in:

- https://github.com/erigontech/erigon/pull/23889#discussion_r3990208781
- https://github.com/erigontech/erigon/pull/23889#discussion_r3990208791
- https://github.com/erigontech/erigon/pull/23889#discussion_r3990208798

## Changes

- move the Gloas payload-attestation poll into a locally testable script
- distinguish candidate scan gaps, head API failures, chain liveness failures, and complete scans without an available payload
- retain and retry unresolved candidate slots while continuing to inspect later slots
- retry near-head 404s that can occur before the numeric canonical index catches up
- reject malformed, multi-document, wrong-version, wrong-slot, or invalid-attestation HTTP 200 responses
- require canonical unsigned-decimal slot strings and compare them exactly without lossy JSON-number conversion
- reject head slots that could overflow Bash arithmetic
- reject initial-head, later-head, and candidate results if request or bounded parsing completes after the poll deadline
- cap Beacon API response bodies at 8 MiB so post-request parsing work is bounded
- make the missing Gloas bid test actually pass a nil bid and assert the default body value
- run the shell test from the lint workflow

## Verification

- `make test-kurtosis-gloas-poll` (three consecutive runs)
- `go test ./cmd/caplin/caplin1 -run '^TestDevGenesisBeaconBody' -count=1`
- `make lint` (two consecutive runs)
- `make erigon integration`
- nil-guard mutation check: the corrected test fails when the production nil guard is removed and passes after restoration
- deadline-guard evidence check: a valid candidate whose parsing crosses the deadline cannot report success
- independent canonical-representation, boundary/lifetime, and evidence-truth reviews converged with no remaining actionable findings

The full Kurtosis Gloas workflow is left to this PR's CI run.

### CODE REVIEW

File: .github/workflows/scripts/check_gloas_payload_attestations.test.sh [hunk 1]
```diff
@@ -0,0 +1,495 @@
...
+if output=$(GLOAS_POLL_TIMEOUT_SECONDS=1 GLOAS_POLL_TARGET_ADVANCE=1 GLOAS_POLL_SLEEP_SECONDS=0 \
+  "$poll_script" http://beacon.test "$fixture_dir/api" 2>&1); then
+  echo "expected candidate parsed after the deadline to fail" >&2
+  exit 1
+fi
+grep -q "Payload-attestation scan incomplete before the deadline" <<<"$output"
+if grep -q "Found available payload attestations" <<<"$output"; then
+  echo "candidate parsed after the deadline was reported as successful: $output" >&2
+  exit 1
+fi
```
Comment (awskii) [inline .github/workflows/scripts/check_gloas_payload_attestations.test.sh:404]: **late_candidate_parse scenario is timing-flaky with a 1-second SECONDS deadline**

The scenario runs the poll script with `GLOAS_POLL_TIMEOUT_SECONDS=1` and expects `Payload-attestation scan incomplete`. The test relies on the deadline passing only during the jq wrapper's `sleep` on the first candidate parse. Bash `SECONDS` counts whole wall-clock seconds, so `deadline=$((SECONDS + 1))` can expire anywhere from about 0 s to 1 s after it is set. In local runs it expired after 0.111 s, 0.637 s and 0.883 s.

Before the first candidate jq call at script line 151, the script runs the `normalize_poll_config` subshells, the initial head fetch through fake curl plus its jq parse, then a second head fetch and parse. If the second boundary falls inside that window, the run takes a different path:
- during the initial fetch, line 78 reports `no valid head slot was observed`
- before line 100, the main loop never runs, and the output is `Chain liveness failure` (head 0 < target 1)
- during the later head fetch, line 117 breaks, again giving `Chain liveness failure`

In each case `grep -q "Payload-attestation scan incomplete"` fails, and `set -e` aborts the test. The window is tens of milliseconds of fork/exec, so the test fails on a few percent of runs. It now runs in `lint.yml` (`make test-kurtosis-gloas-poll`), so it will fail the lint job at random on unrelated PRs.

Fix: Make the timing deterministic: leave more than one guaranteed second before the first candidate parse, and make the jq delay longer than the timeout. For example, use `GLOAS_POLL_TIMEOUT_SECONDS=2` with a 3 s sleep, or `3` with a 4 s sleep, in the jq wrapper for this scenario.

File: .github/workflows/scripts/check_gloas_payload_attestations.sh [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+            all(.data.message.body.payload_attestations[];
+              .data.slot == $attestation_slot and
+              (.aggregation_bits |
+                test("^0x[0-9a-fA-F]+$") and
+                (test("^0x0+$"; "i") | not)
+              ) and
+              (.signature |
+                test("^0x[0-9a-fA-F]{192}$") and
+                (test("^0x0+$"; "i") | not)
+              )
```
Comment (awskii) [inline .github/workflows/scripts/check_gloas_payload_attestations.sh:190]: **Payload-attestation validation skips beacon_block_root and blob_data_available and accepts any aggregation_bits width**

The structural jq (lines 157-172) never reads `.data.beacon_block_root` or `.data.blob_data_available`. Both fields are part of `PayloadAttestationData` (`cl/cltypes/epbs_payload.go:78-81`). `aggregation_bits` only has to match `^0x[0-9a-fA-F]+$` and not be all zeros (lines 183-186). `BitVector.MarshalJSON` emits the full container as hex, and the `gloas-caplin-genesis` suite uses `preset: mainnet` (PTC_SIZE 512), so a valid value has exactly 128 hex digits. The fake-curl fixture (test.sh:220) emits `0x01` and omits both data fields, and it still reaches the success path, so the tests encode the malformed shape as canonical.

Impact: if a Caplin Beacon API JSON regression dropped these fields or truncated the bitvector, this Gloas CI check would still pass. Caplin serializes these fields from Go structs, so that regression is unlikely, and the check mainly guards liveness. The old inline workflow had the same gaps. This PR explicitly claims to reject invalid-attestation HTTP 200 responses and rewrites the validator, so the gap is in scope. The matching Copilot threads are still unresolved.

Fix: In the structural jq, add `(.data.beacon_block_root | type == "string" and test("^0x[0-9a-fA-F]{64}$"))` and `(.data.blob_data_available | type == "boolean")`. Require `aggregation_bits` to match `^0x[0-9a-fA-F]{128}$`, or take the width from `PTC_SIZE` in `/eth/v1/config/spec`. Update the fixtures to emit these fields at the correct width.

File: .github/workflows/scripts/check_gloas_payload_attestations.sh [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+  echo "Waiting for payload attestations: head slot $head_slot, target slot $target_head_slot"
+  wait_for_next_poll || break
+done
+
+if [ "$scan_incomplete" = true ]; then
+  echo "::error::Payload-attestation scan incomplete before the deadline; last head slot $last_head_slot, next unverified slot $next_slot"
+elif [ "$head_fetch_unavailable" = true ]; then
+  echo "::error::Head polling was incomplete before the deadline; last observed head slot $last_head_slot"
+elif ((last_head_slot < target_head_slot)); then
+  echo "::error::Chain liveness failure: head only advanced from slot $initial_head_slot to $last_head_slot before the deadline"
```
Comment (awskii) [inline .github/workflows/scripts/check_gloas_payload_attestations.sh:251]: **A stalled chain with a missed near-head slot is reported as an incomplete scan instead of a liveness failure**

Suppose the head stops at slot H and slot H-1 was missed. Then `blocks/H-1` returns 404 on every poll. While the head does not move, `slot + 1 >= head_slot` stays true, so the 404 branch marks the slot unresolved on every pass, and `scan_incomplete` stays true until the deadline. The final classifier checks `scan_incomplete` (line 246) before `last_head_slot < target_head_slot` (line 250), so the step prints `Payload-attestation scan incomplete ... next unverified slot H-1` instead of `Chain liveness failure`.

Trace: initial head 0, target 8, later head stuck at 2, slot 1 returns 404. Every pass marks slot 1 unresolved, sets `next_slot=1` and waits. At the deadline the scan-incomplete message is printed. The step still fails, but it names the wrong cause and points the investigation at the API index rather than chain liveness. That defeats the point of telling scan gaps and liveness failures apart. AskAlexSharov reported this against dd02161a57b, and 4625a133ee2 did not change the logic.

Fix: Determine the terminal head state before choosing the diagnostic. When a successfully observed head stays below target and the only unresolved slots are near-head 404s, report liveness first. For example, track a `near_head_404_only` flag, or check `last_head_slot < target_head_slot` before `scan_incomplete` in that case.

File: .github/workflows/scripts/check_gloas_payload_attestations.test.sh [hunk 4]
```diff
@@ -0,0 +1,495 @@
...
+  esac
+done
+
+if [[ "$url" == */head ]]; then
+  count_file="$GLOAS_TEST_STATE_DIR/head-count"
+  count=$(($(cat "$count_file" 2>/dev/null || echo 0) + 1))
+  echo "$count" >"$count_file"
+  if [[ "$GLOAS_TEST_SCENARIO" == initial_deadline && "$max_time" -gt "$GLOAS_POLL_TIMEOUT_SECONDS" ]]; then
+    exit 87
+  fi
```
Comment (awskii) [inline .github/workflows/scripts/check_gloas_payload_attestations.test.sh:56]: **initial_deadline test does not notice if --max-time is removed from the initial head fetch**

The fake curl rejects the request only when `[[ "$max_time" -gt "$GLOAS_POLL_TIMEOUT_SECONDS" ]]`. `max_time` starts empty (line 19), and inside `[[ ]]` an empty operand to `-gt` counts as 0, so the check is false when the flag is missing. If someone deletes `--max-time "$request_timeout"` from the initial head request (script lines 74-77), the fake curl returns a normal response and the scenario still reports `Found available payload attestations`. `initial_late_response` also ignores `max_time`. An unbounded first request would therefore pass CI, although the initial fetch is meant to be bounded by the deadline. The `--max-filesize` check next to it does catch a missing flag.

Fix: Treat a missing flag as a failure: `if [[ "$GLOAS_TEST_SCENARIO" == initial_deadline && ( -z "$max_time" || "$max_time" -gt "$GLOAS_POLL_TIMEOUT_SECONDS" ) ]]; then exit 87; fi`.

File: .github/workflows/scripts/check_gloas_payload_attestations.test.sh [hunk 5]
```diff
@@ -0,0 +1,495 @@
...
+    printf '{"data":{"message":{"slot":"01"}}}\n' >"$output"
+    exit 0
+  fi
+  if [[ "$GLOAS_TEST_SCENARIO" == noncanonical_head ]] && ((count == 3)); then
+    printf '{"data":{"message":{"slot":"1\\n"}}}\n' >"$output"
+    exit 0
+  fi
+  if [[ "$GLOAS_TEST_SCENARIO" == multi_document_head ]] && ((count == 2)); then
+    printf '{}\n{"data":{"message":{"slot":"1"}}}\n' >"$output"
+    exit 0
```
Comment (awskii) [inline .github/workflows/scripts/check_gloas_payload_attestations.test.sh:97]: **Multi-document tests still pass when the single-document guard is removed**

Both multi-document fixtures send the invalid `{}` document first: the head fixture (lines 95-96) and the candidate fixture (lines 181-183). With `jq -s`, `.[0]` is `{}`, which fails the later field checks by itself, so the response is rejected even without `select(length == 1)`. Running the filters through jq with the guard removed: On the fixture order, `read_head_slot` exits 4 (rejected, test passes). On the reversed order, with the valid document first, it prints `1` and exits 0. So `multi_document_head` and `multi_document_candidate` stay green if the guards at script lines 58 and 157 are deleted, and a regression that accepts a body with trailing documents would pass.

Fix: Put the valid document first and the extra document after it, e.g. `printf '{"data":{"message":{"slot":"1"}}}\n{}\n'` for the head and the valid block followed by `{}` for the candidate.

File: .github/workflows/scripts/check_gloas_payload_attestations.test.sh [hunk 6]
```diff
@@ -0,0 +1,495 @@
...
+grep -q "Payload-attestation scan incomplete" <<<"$output"
+if grep -Eq "Chain liveness failure|No canonical Gloas block" <<<"$output"; then
+  echo "deadline scenario used a misleading diagnosis: $output" >&2
+  exit 1
+fi
+
+cleanup
+new_fixture
+export GLOAS_TEST_SCENARIO=retry
+output=$(GLOAS_POLL_TIMEOUT_SECONDS=30 GLOAS_POLL_TARGET_ADVANCE=1 GLOAS_POLL_SLEEP_SECONDS=0 \
```
Comment (awskii) [inline .github/workflows/scripts/check_gloas_payload_attestations.test.sh:263]: **Test harness fails silently: bare set -e assertions give no scenario name and discard the script output**

The test uses `set -euo pipefail` with bare assertions. In a success scenario, `output=$(... "$poll_script" ...)` aborts the harness when the script exits 1, and the captured `::error::` diagnosis is never printed. A failing `grep -q ... <<<"$output"` or `test "$(cat ...)" -eq N` likewise exits 1 with no message. In CI, `make test-kurtosis-gloas-poll` shows only `make: *** Error 1`, with nothing to say which of the ~25 scenarios failed or why. Other harnesses in the same directory avoid this: `ci-gate-check.test.sh:13-51` has `run_case`/`run_output` with `FAIL - <name>` output and a pass/fail summary, and `check-large-files.test.sh` and `setup-kurtosis/setup.test.sh` report failures by name too. The new file also repeats the same 3-line invocation block for every scenario.

Fix: Add a `run_scenario <name> <want_exit> <want_regex> [count-file expected]` helper modelled on `run_output` in ci-gate-check.test.sh. It should capture the exit code without `set -e`, print `FAIL - <name>` with the output on mismatch, count passes and failures, and end with `[ "$fail" -eq 0 ]`.

File: .github/workflows/scripts/check_gloas_payload_attestations.sh [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+                (test("^0x0+$"; "i") | not)
+              ) and
+              (.signature |
+                test("^0x[0-9a-fA-F]{192}$") and
+                (test("^0x0+$"; "i") | not)
+              )
+            )
+          ' "$api_tmp_dir/candidate.json"; then
+            echo "Candidate block response for slot $slot was invalid; retrying"
+            slot_unresolved=true
```
Comment (awskii) [inline .github/workflows/scripts/check_gloas_payload_attestations.sh:194]: **Candidate validation is split into two overlapping jq passes that take the same retry branch**

Each HTTP 200 candidate is parsed by three jq processes: a structural pass (151-173), a semantic pass (176-192) and the `any(... payload_present == true)` check (195-199). The first two go to the same `Candidate block response ... was invalid; retrying` branch and mostly overlap. Candidate-side `canonical_decimal` on `.data.message.slot` and `.data.slot` is implied by the exact string equality with the bash-generated `$requested_slot`/`$attestation_slot`. The `type == "string"` guards on `aggregation_bits`/`signature` are redundant, because `test()` on a non-string raises an error (jq -e exit 5) and takes the same branch.

The structural pass is not fully redundant, though. Besides `select(length == 1)`, it is the only place that requires `payload_present` to be a boolean. Without it, a string `"true"` resolves the slot as having no attestation instead of retrying. It also requires `payload_attestations` to be an array: `.[]` on an object would iterate its values. A maintainer asked to remove the duplicate type-check pass on an earlier revision. With the current split, a change to a validation rule has to be made in two filters.

Fix: Merge lines 151-192 into one `jq -se --arg requested_slot … --arg attestation_slot …` filter: `select(length == 1) | .[0] |`, then the object/array type guards, the `payload_present | type == "boolean"` check, and the exact-equality and hex checks. Drop the candidate-side `canonical_decimal` definition and the redundant string-type guards.

**Review by awskii — CHANGES_REQUESTED** [review]
Requesting changes for the timing-flaky `late_candidate_parse` scenario, which now runs in the lint job. Six minor points inline.

**Comment by domiwei** [comment]
Pushed 9a64b949ae to address the latest validation/deadline feedback: required PayloadAttestationData fields are checked; mainnet aggregation bits require 128 canonical hex digits; roots/signatures use absolute-end validation; attestation roots are linked to the block parent root with case-normalized Bytes32 equality; late head and both candidate availability outcomes are classified after the deadline; and the transient-head recovery/liveness scenarios are deterministic. The shell suite passed three consecutive runs, focused Caplin test and builds pass, lint passed repeatedly, and independent lifecycle, boundary, and evidence-truth reviews converged.

File: .github/workflows/scripts/check_gloas_payload_attestations.sh (hunk 3 above)
Comment (domiwei) [reply .github/workflows/scripts/check_gloas_payload_attestations.sh:333]: Fixed across c93e5c88 and d641595fe2. The poller tracks whether unresolved slots are exclusively near-head 404s, reserves a terminal head refresh, and reports liveness when the refreshed head remains below target. The `stalled chain with near-head 404 reports liveness` test covers the stated head 0 -> 2 / slot 1 = 404 sequence, while separate tests preserve scan-incomplete for real API holes and head-fetch failures.

File: .github/workflows/scripts/check_gloas_payload_attestations.sh (hunk 7 above)
Comment (domiwei) [reply .github/workflows/scripts/check_gloas_payload_attestations.sh]: Fixed in c93e5c88 and hardened further in d641595fe2. Candidate validation and `payload_present` evaluation are now one jq program, with single-document, exact slot/version, required-field, fixed-width, root-equality, signature, and mainnet max-cardinality checks before `any(...)`. The jq process is also bounded by the scan deadline so it cannot consume the terminal head-refresh budget.

File: .github/workflows/scripts/check_gloas_payload_attestations.test.sh (hunk 4 above)
Comment (domiwei) [reply .github/workflows/scripts/check_gloas_payload_attestations.test.sh]: Fixed in c93e5c88. The fake curl now rejects a missing `--max-time` as well as a value above the configured timeout, so removing the initial request bound fails `initial request obeys the poll deadline`. d641595fe2 additionally verifies candidate requests reserve the terminal two-second head budget.

File: .github/workflows/scripts/check_gloas_payload_attestations.test.sh (hunk 6 above)
Comment (domiwei) [reply .github/workflows/scripts/check_gloas_payload_attestations.test.sh]: Fixed in c93e5c88. The harness now uses a named `run_scenario` helper, captures exit/output without aborting, prints `FAIL - <scenario>` plus the script output and assertion detail, and finishes with a pass/fail summary. The latest run reports 52 passed, 0 failed.

File: .github/workflows/scripts/check_gloas_payload_attestations.test.sh (hunk 5 above)
Comment (domiwei) [reply .github/workflows/scripts/check_gloas_payload_attestations.test.sh]: Fixed in c93e5c88. Both multi-document fixtures now emit the valid document first and the extra `{}` second, so deleting `select(length == 1)` would accept the first document and fail the request-count assertions. The full shell harness passes 52/52.

**Comment by domiwei** [comment]
Pushed d641595fe2 to close the remaining review gaps:

- enforce exact HTTP 200 for initial and later head samples;
- cap mainnet payload attestations at 4 and reject mixed malformed arrays;
- reserve a terminal head sample so stale or post-deadline state cannot be classified as liveness evidence;
- bound both candidate curl and jq work by the scan deadline;
- make the near-head 404 -> valid empty-block transition deterministic;
- strengthen the deadline tests to prove the terminal head request actually occurred.

Validation: the shell harness passes 52/52 repeatedly, make lint passes repeatedly, and bash -n plus git diff --check are clean. Independent boundary, lifecycle, and evidence-truth reviews converged. I replied to and resolved the five remaining review threads individually.
